pkg/proxy/ipvs/proxier.go

```diff
@@ -647,29 +641,31 @@ func (handle *LinuxKernelHandler) GetModules() ([]string, error) {
 		return nil, fmt.Errorf("failed to find loaded kernel modules: %v", err)
 	}
 
-	builtinModsFilePath := fmt.Sprintf("/lib/modules/%s/modules.builtin", kernelVersionStr)
-	b, err := ioutil.ReadFile(builtinModsFilePath)
+	bmods, nfconntrack, err = utilipvs.BuiltinIPVSModules(false, kernelVersionStr)
 	if err != nil {
-		klog.Warningf("Failed to read file %s with error %v. You can ignore this message when kube-proxy is running inside container without mounting /lib/modules", builtinModsFilePath, err)
+		return nil, fmt.Errorf("Failed to get kernel built-in modules")
 	}
 
+	mods = append(mods, bmods...)
 	for _, module := range ipvsModules {
-		if match, _ := regexp.Match(module+".ko", b); match {
-			bmods = append(bmods, module)
+		// Try to load the required IPVS kernel modules if not built in
+		err := handle.executor.Command("modprobe", "--", module).Run()
+		if err != nil {
+			klog.Warningf("Failed to load kernel module %v with modprobe. "+
+				"You can ignore this message when kube-proxy is running inside container without mounting /lib/modules or if the module is built-in", module)
 		} else {
-			// Try to load the required IPVS kernel modules if not built in
-			err := handle.executor.Command("modprobe", "--", module).Run()
-			if err != nil {
-				klog.Warningf("Failed to load kernel module %v with modprobe. "+
-					"You can ignore this message when kube-proxy is running inside container without mounting /lib/modules", module)
-			} else {
-				lmods = append(lmods, module)
+			if module == utilipvs.KernelModuleNfConntrack ||
+				module == utilipvs.KernelModuleNfConntrackIPV4 {
+				nfconntrack = true
```

Comment:
You're hardcoding in two different places the fact that all of the modules returned by `GetRequiredIPVSModules` are required, _except_ for those two, which you only need one of... that's a confusing API.

If you want to be really abstract, `GetRequiredIPVSModules` could return a `[][]string`, where each element of the outer array is an array of equivalent modules, and the caller would make sure that at least one module from each element was present:

```
	return [][]string{
		[]string{KernelModuleIPVS},
		[]string{KernelModuleIPVSRR},
		[]string{KernelModuleIPVSWRR},
		[]string{KernelModuleIPVSSH},
		[]string{KernelModuleNfConntrack, KernelModuleNfConntrackIPV4},
	}
```

 But maybe that's excessive? Maybe just have it return an `allof` array and a `oneof` array?